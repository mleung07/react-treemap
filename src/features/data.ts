import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "../app/store";

export interface Item {
  name: string;
  weight: number;
  value: number;
}

export interface TreeState {
  tree: Item[];
  rows: number;
}

const initialState: TreeState = {
  tree: [],
  rows: 1,
};

export const sampleTree: Item[] = [
  { name: "A", weight: 3, value: -0.02 },
  { name: "B", weight: 3, value: 0.05 },
  { name: "C", weight: 6, value: 0.015 },
  { name: "D", weight: 2, value: -0.01 },
  { name: "E", weight: 3, value: 0.01 },
];

export const defaultRows = 3;

export const dataSlice = createSlice({
  name: "data",
  initialState,
  reducers: {
    initData: (state) => {
      state.tree = sampleTree;
      state.rows = defaultRows;
    },
    updateTree: (state, action: PayloadAction<Item[]>) => {
      state.tree = action.payload;
      state.rows = Math.max(1, Math.min(state.rows, action.payload.length));
    },
    updateRows: (state, action: PayloadAction<number>) => {
      state.rows = action.payload;
    },
  },
});

export const { initData, updateTree, updateRows } = dataSlice.actions;

// selectors
export const selectTree = (state: RootState) => state.tree;
export const selectRows = (state: RootState) => state.rows;
export const selectItemSize = (state: RootState) => state.tree.length;
export const selectMaxWeightPerRow = (state: RootState) => {
  // calculation: total weight / no. of rows
  // if any of the item has a larger width, update the max weight
  const maxWeight = Math.max(...state.tree.map((item) => item.weight));
  const totalWeight = state.tree.reduce((total, current) => {
    return total + current.weight;
  }, 0);
  return Math.max(maxWeight, Math.ceil(totalWeight / state.rows));
};

export default dataSlice.reducer;
