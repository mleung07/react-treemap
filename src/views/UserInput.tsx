import TreeInput from "../components/TreeInput";
import RowInput from "../components/RowInput";

const UserInput = () => {
  return (
    <div>
      <div>
        <h3>Data</h3>
        <TreeInput />
      </div>
      <div>
        <h3>Row Number</h3>
        <RowInput />
      </div>
    </div>
  );
};

export default UserInput;
