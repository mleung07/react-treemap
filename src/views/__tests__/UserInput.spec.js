import React from "react";
import { render } from "@testing-library/react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import UserInput from "../UserInput";
import { sampleTree } from "../../features/data";

const mockStore = configureStore([]);

it("should render UserInput correctly", () => {
  const initialState = { rows: 3, tree: sampleTree };
  const store = mockStore(initialState);

  const elem = render(
    <Provider store={store}>
      <UserInput />
    </Provider>
  );
  expect(elem).toMatchSnapshot();
});
