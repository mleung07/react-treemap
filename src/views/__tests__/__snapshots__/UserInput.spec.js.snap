// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserInput correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <h3>
            Data
          </h3>
          <div>
            <textarea
              aria-label="tree-input"
              class="sc-bdfBQB gSvzUv"
              cols="50"
              rows="20"
            >
              [
  {"name": "A", "weight": 3, "value": -0.02},
  {"name": "B", "weight": 3, "value": 0.05},
  {"name": "C", "weight": 6, "value": 0.015},
  {"name": "D", "weight": 2, "value": -0.01},
  {"name": "E", "weight": 3, "value": 0.01}
]
            </textarea>
            <div
              class="sc-gsTEea hxmvRm"
            />
          </div>
        </div>
        <div>
          <h3>
            Row Number
          </h3>
          <div>
            <input
              aria-label="rows-input"
              class="sc-dlfnuX isBgPT"
              max="5"
              min="1"
              type="number"
              value="3"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <h3>
          Data
        </h3>
        <div>
          <textarea
            aria-label="tree-input"
            class="sc-bdfBQB gSvzUv"
            cols="50"
            rows="20"
          >
            [
  {"name": "A", "weight": 3, "value": -0.02},
  {"name": "B", "weight": 3, "value": 0.05},
  {"name": "C", "weight": 6, "value": 0.015},
  {"name": "D", "weight": 2, "value": -0.01},
  {"name": "E", "weight": 3, "value": 0.01}
]
          </textarea>
          <div
            class="sc-gsTEea hxmvRm"
          />
        </div>
      </div>
      <div>
        <h3>
          Row Number
        </h3>
        <div>
          <input
            aria-label="rows-input"
            class="sc-dlfnuX isBgPT"
            max="5"
            min="1"
            type="number"
            value="3"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
